promql/engine.go

```diff
@@ -1325,16 +1380,101 @@ func (ev *evaluator) eval(expr parser.Expr) (parser.Value, storage.Warnings) {
+		subqueryStartTime := ev.startTimestamp
+		if subqueryEvalTime != 0 {
+			if subqueryEvalTime > ev.endTimestamp {
+				newEv.endTimestamp = subqueryEvalTime - offsetMillis
+			}
+			if subqueryEvalTime < subqueryStartTime {
+				subqueryStartTime = subqueryEvalTime
+			}
+		}
+
+		// Start with the first timestamp after (subqueryStartTime - offset - range)
 		// that is aligned with the step (multiple of 'newEv.interval').
-		newEv.startTimestamp = newEv.interval * ((ev.startTimestamp - offsetMillis - rangeMillis) / newEv.interval)
-		if newEv.startTimestamp < (ev.startTimestamp - offsetMillis - rangeMillis) {
+		newEv.startTimestamp = newEv.interval * ((subqueryStartTime - offsetMillis - rangeMillis) / newEv.interval)
+		if newEv.startTimestamp < (subqueryStartTime - offsetMillis - rangeMillis) {
 			newEv.startTimestamp += newEv.interval
 		}
 
 		res, ws := newEv.eval(e.Expr)
 		ev.currentSamples = newEv.currentSamples
 		return res, ws
+	case *parser.ConstantExpr:
+		switch ce := e.Expr.(type) {
+		case *parser.StringLiteral, *parser.NumberLiteral:
+			return ev.eval(ce)
+		}
+
+		if e.Result != nil {
+			return e.Result, nil
+		}
+		newEv := &evaluator{
+			startTimestamp:           ev.startTimestamp,
+			endTimestamp:             ev.startTimestamp,
+			interval:                 ev.interval,
+			ctx:                      ev.ctx,
+			currentSamples:           ev.currentSamples,
+			maxSamples:               ev.maxSamples,
+			logger:                   ev.logger,
+			lookbackDelta:            ev.lookbackDelta,
+			noStepSubqueryIntervalFn: ev.noStepSubqueryIntervalFn,
+		}
+
+		canExtrapolate := canExtrapolateConstantExpr(e.Expr)
+		if !canExtrapolate {
+			newEv.endTimestamp = ev.endTimestamp
+		}
+
+		res, ws := newEv.eval(e.Expr)
+		ev.currentSamples = newEv.currentSamples
+
+		if canExtrapolate {
+			switch val := res.(type) {
+			case Matrix:
+				for i := range val {
+					if len(val[i].Points) != 1 {
+						panic(errors.Errorf("unexpected number of samples"))
+					}
+					for ts := ev.startTimestamp + ev.interval; ts <= ev.endTimestamp; ts = ts + ev.interval {
+						val[i].Points = append(val[i].Points, Point{
+							T: ts,
+							V: val[i].Points[0].V,
+						})
+						ev.currentSamples++
+						if ev.currentSamples >= ev.maxSamples {
+							ev.error(ErrTooManySamples(env))
+						}
+					}
+				}
+			case Vector:
+				mat := make(Matrix, len(val))
+				for i, smpl := range val {
+					mat[i] = Series{
+						Metric: smpl.Metric,
+						Points: make([]Point, 0, 1+(ev.endTimestamp-ev.startTimestamp)/ev.interval),
+					}
+					for ts := ev.startTimestamp + ev.interval; ts <= ev.endTimestamp; ts = ts + ev.interval {
+						mat[i].Points = append(mat[i].Points, Point{
+							T: ts,
+							V: smpl.V,
+						})
+						ev.currentSamples++
+						if ev.currentSamples >= ev.maxSamples {
+							ev.error(ErrTooManySamples(env))
+						}
+					}
+				}
+				res = mat
+			default:
+				panic(errors.Errorf("unexpected result: %T", expr))
```

Comment:
I will add a case for that, but as I see, we should not extrapolate scalar, is that right? Or does the timestamp of the scalar matter and would change and hence we need to extrapolate?

Reply:
The extrapolation logic shouldn't exist, but aside from that expressions can return scalars so you need to deal with that.

Reply:
> The extrapolation logic shouldn't exist

Even for Vector and Matrix? That would be very inefficient

Reply:
I mean the canExtrapolate stuff. Nothing in this PR should be doing anything relating to extrapolation.

Reply:
With extrapolate I mean duplicate the results to other evaluations (calculate for 1 step and replicate the result for other steps instead of calculating it again). I guess that should be fine? (I can rename some things around here)